### ISSUE TITLE
Installed qt5 event loop hook.

### ISSUE THREAD

**Issue by forallsunday** [ipython/ipython#14006]
When I use `plt.show()` , and run the script,
``` bash
❯ & D:/Anaconda3/python.exe -m IPython --no-autoindent d:/Documents/C-Project/GeoLocOptim/scripts/estimate.py
```

The following message showed:
```
Installed qt5 event loop hook.
Shell is already running a gui event loop for qt5. Call with no arguments to disable the current loop.
```

What does it mean and how to disable this?

**Comment by nodatapoints** [comment]
I think this is due to #13957.

**Comment by francescoboc** [comment]
I agree with @nodatapoints , I would like to suppress the message `Installed qt5 event loop hook.`.

**Comment by ccordoba12** [comment]
@shaperilio, how do you think we should address this issue?

**Comment by shaperilio** [comment]
Is there a global "debug" flag?

The print statements were added precisely because silent failures were happening before; i.e. you run something which is forcing a `qt` event loop hook without asking, and then you muck around with some `tk` stuff and nothing works.

Happy to help make this quieter, and maybe work on something that allows multiple event loop hooks to run in sequence, which I believe should be possible as well.

**Comment by ccordoba12** [comment]
I think the problem is that Matplotlib sets the Qt backend by default in some cases, which generates this message in IPython when users try to set the backend again.

So, perhaps the message shouldn't be shown when the backend you're trying to set is the same one that you already have.

**Comment by shaperilio** [comment]
> So, perhaps the message shouldn't be shown when the backend you're trying to set is the same one that you already have.

This was certainly the intent. Let me take a look at what's going on.

@forallsunday can you post relevant package versions?

**Comment by nodatapoints** [comment]
> @forallsunday can you post relevant package versions?

I run IPython 8.12.0 with Qt5 on Gentoo. In my configuration I set
```c.TerminalIPythonApp.matplotlib = 'qt5'```
otherwise I use default settings except for some prompt configs.

**Comment by shaperilio** [comment]
> I run IPython 8.12.0 with Qt5 on Gentoo. In my configuration I set `c.TerminalIPythonApp.matplotlib = 'qt5'` otherwise I use default settings except for some prompt configs.

What is `c`?

**Comment by ccordoba12** [comment]
It's the IPython config object. See here for more details:

https://ipython.readthedocs.io/en/stable/config/intro.html

**Comment by ccordoba12** [comment]
I think ignoring warnings for the same event loop is fine for now.

**Comment by shaperilio** [comment]
@tacaswell do you have any thoughts on the backend<->GUI mapping?

**Comment by tacaswell** [comment]
There is a winding road to get here and I am not sure I have all of the details correct, but my rough understanding is:

 - IPython provides nicer names (without the agg) to select the backend
 - these names date back probably 15+ years and the current versions of Qt
 - in Matplotlib we had a pattern of doing a new backend for each GUI toolkit version (backend_qt4agg.py, backend_qt5agg.py, backend_gtk3agg.py, backend_gtk4agg.py) etc
 - when qt6 came out instead of making a `backend_qt6agg.py` we managed to push enough compatibility shims into `backends/qt_compat.py` that we could support all 4 Qt bindings (pyside / pyqt X Qt5, Qt6) and made one new backend called `backend_qtagg.py`. https://github.com/matplotlib/matplotlib/pull/19255 .  This is mpl3.5
    - as part of this we also dropped almost all of our machinery for picking the import you want...if you care import your preferred binding before us! See https://matplotlib.org/stable/api/backend_qt_api.html#qt-bindings
 - we kept `backend_qt5agg.py` around as a very thin shim (and a knob to force `qt_compat.py` to pick one of the Qt5 bindings) because there is no good reason to break any users who are importing directly
 - It looks like I was the last person to touch the forward mapping of `'qt'` in 127b45b9866f2b9dcf0a1f880eebcf554a79ee5c when I changed `'qt'` to map to qt5agg!
 - There is a bug in the table at https://matplotlib.org/stable/users/explain/backends.html#the-builtin-backends which claims `%matplotlib qt` selects `qtagg` 

I think the course of action here should be:

- version gate the meaning of `qt` to be `qtagg`

I'll open this PR.

This is closely related to https://github.com/ipython/ipython/issues/14038

**Comment by shaperilio** [comment]
Thanks @tacaswell; I'll need some time to process your comment.

In the meantime, I've started a draft pull request. The messaging is improved (that's the easy part) but I left a comment there because importing `matplotlib_inline.backend_inline` triggers a GUI event loop hook that I don't think is needed, but I need more guidance on that.

**Comment by tacaswell** [comment]
> but I left a comment there because importing matplotlib_inline.backend_inline triggers a GUI event loop hook that I don't think is needed, but I need more guidance on that.

I'm 75% sure that is something that needs to be resolved in the inline backend.

**Comment by tacaswell** [comment]
The other thing to check is that the default backend in Matplotlib in now "auto", however if you ask Matplotlib what the current backend is or create a figure we will sort out what the best currently available backend is.  If you have any of the Qt bindings installed, we will pick that (as interactive figures are better than dead figures).

https://matplotlib.org/stable/users/explain/backends.html#selecting-a-backend

> Without a backend explicitly set, Matplotlib automatically detects a usable backend based on what is available on your system and on whether a GUI event loop is already running. The first usable backend in the following list is selected: MacOSX, QtAgg, GTK4Agg, Gtk3Agg, TkAgg, WxAgg, Agg. The last, Agg, is a non-interactive backend that can only write to files. It is used on Linux, if Matplotlib cannot connect to either an X display or a Wayland display.

I do not see `matplotlib_inline` doing anything I expect to force the backend, but I may have missed it or am forgetting another condition that force us to resolve the backend.

**Comment by shaperilio** [comment]
> The other thing to check is that the default backend in Matplotlib in now "auto"

This is exactly what `matlplotlib_inline` is triggering, by calling [`_enable_matplotlib_integration()`](https://github.com/ipython/matplotlib-inline/blob/fbf0ab8f2d81993ff8e319a003f58896c77f2443/matplotlib_inline/backend_inline.py#L228) in `backend_inline.py`. That calls `matplotlib.get_backend()`, which will trigger `__getitem__` in `RcParams` and execute this:

https://github.com/matplotlib/matplotlib/blob/5189609b9d2c0ea0b78ceba9d4003358db160b81/lib/matplotlib/__init__.py#L750-L754

and that last line will go through the list mentioned in the docs.

So yes, it seems that `_enable_matplotlib_integration()` should probably call `matplotlib._get_backend_or_none()`.

**Comment by shaperilio** [comment]
OK, #14067 implements what I think we should do.

I think it needs to wait to see what we hear from [`matplotlib_inline`](https://github.com/ipython/matplotlib-inline/issues/25).

**Comment by shaperilio** [comment]
I just wrote up #14089; maybe we can take the choice away from users who don't really care? Especially with this Qt version stuff, I wonder if we support it just because the code is there and not because most people actually need it.
